File: Python/Multiply_two_matrix.py
def multiply_matrices(
    matrix_a: list[list[float]], matrix_b: list[list[float]]
) -> list[list[float]]:
    if not matrix_a or not matrix_b:
        raise ValueError("Matrices cannot be empty.")

    rows_a = len(matrix_a)
    cols_a = len(matrix_a[0])
    rows_b = len(matrix_b)
    cols_b = len(matrix_b[0])

    for row in matrix_a:
        if len(row) != cols_a:
            raise ValueError("Matrix A has inconsistent row sizes.")

    for row in matrix_b:
        if len(row) != cols_b:
            raise ValueError("Matrix B has inconsistent row sizes.")

    if cols_a != rows_b:
        raise ValueError(
            f"Cannot multiply: columns of A ({cols_a}) must equal rows of B ({rows_b})."
        )

    result = [[0.0 for _ in range(cols_b)] for _ in range(rows_a)]

    for i in range(rows_a):
        for j in range(cols_b):
            for k in range(cols_a):
                result[i][j] += matrix_a[i][k] * matrix_b[k][j]

    return result


def read_matrix(name: str) -> list[list[float]]:
    rows = int(input(f"Enter number of rows for {name}: ").strip())
    cols = int(input(f"Enter number of columns for {name}: ").strip())

    matrix: list[list[float]] = []
    print(f"Enter rows for {name} (space-separated numbers):")

    for r in range(rows):
        row_values = input(f"Row {r + 1}: ").strip().split()
        if len(row_values) != cols:
            raise ValueError(
                f"Expected {cols} values in row {r + 1}, got {len(row_values)}."
            )
        matrix.append([float(value) for value in row_values])

    return matrix


def print_matrix(matrix: list[list[float]]) -> None:
    for row in matrix:
        print(" ".join(f"{value:g}" for value in row))


def main() -> None:
    try:
        matrix_a = read_matrix("Matrix A")
        matrix_b = read_matrix("Matrix B")
        result = multiply_matrices(matrix_a, matrix_b)

        print("\nResult (A x B):")
        print_matrix(result)
    except ValueError as error:
        print(f"Error: {error}")


if __name__ == "__main__":
    main()

File: Javascript/Multiply_two_matrix.js
const readline = require('node:readline/promises');
const { stdin, stdout } = require('node:process');

const multiplyMatrices = (matrixA, matrixB) => {
  if (!matrixA.length || !matrixB.length) {
    throw new Error('Matrices cannot be empty.');
  }

  const rowsA = matrixA.length;
  const colsA = matrixA[0].length;
  const rowsB = matrixB.length;
  const colsB = matrixB[0].length;

  if (matrixA.some((row) => row.length !== colsA)) {
    throw new Error('Matrix A has inconsistent row sizes.');
  }

  if (matrixB.some((row) => row.length !== colsB)) {
    throw new Error('Matrix B has inconsistent row sizes.');
  }

  if (colsA !== rowsB) {
    throw new Error(
      `Cannot multiply: columns of A (${colsA}) must equal rows of B (${rowsB}).`,
    );
  }

  const result = Array.from({ length: rowsA }, () =>
    Array.from({ length: colsB }, () => 0),
  );

  for (let i = 0; i < rowsA; i += 1) {
    for (let j = 0; j < colsB; j += 1) {
      for (let k = 0; k < colsA; k += 1) {
        result[i][j] += matrixA[i][k] * matrixB[k][j];
      }
    }
  }

  return result;
};

const readMatrix = async (rl, name) => {
  const rows = Number((await rl.question(`Enter number of rows for ${name}: `)).trim());
  const cols = Number((await rl.question(`Enter number of columns for ${name}: `)).trim());

  if (!Number.isInteger(rows) || !Number.isInteger(cols) || rows <= 0 || cols <= 0) {
    throw new Error('Rows and columns must be positive integers.');
  }

  const matrix = [];
  console.log(`Enter rows for ${name} (space-separated numbers):`);

  for (let r = 0; r < rows; r += 1) {
    const rowValues = (await rl.question(`Row ${r + 1}: `))
      .trim()
      .split(/\s+/)
      .filter(Boolean)
      .map(Number);

    if (rowValues.length !== cols || rowValues.some((value) => Number.isNaN(value))) {
      throw new Error(`Expected ${cols} numeric values in row ${r + 1}.`);
    }

    matrix.push(rowValues);
  }

  return matrix;
};

const printMatrix = (matrix) => {
  for (const row of matrix) {
    console.log(row.map((value) => String(value)).join(' '));
  }
};

const main = async () => {
  const rl = readline.createInterface({ input: stdin, output: stdout });

  try {
    const matrixA = await readMatrix(rl, 'Matrix A');
    const matrixB = await readMatrix(rl, 'Matrix B');
    const result = multiplyMatrices(matrixA, matrixB);

    console.log('\nResult (A x B):');
    printMatrix(result);
  } catch (error) {
    console.log(`Error: ${error.message}`);
  } finally {
    rl.close();
  }
};

main();
